frameworks/Swift/perfect/Sources/Perfect-MySQL/main.swift

```diff
@@ -0,0 +1,322 @@
+
+        if let id = row[0], let message = row[1] {
+            
+            let resObj = ["id": String(describing: id), "message": message]
+            arrOfFortunes.append(resObj)
+        } else {
+            print("not correct values returned: ", row)
+        }
+    }
+
+    return arrOfFortunes
+}
+
+func fetchFromWorld(id: String?) -> [String:Any] {
+
+    var returnObj = [String: Any]()
+    var errorObject = [String: Any]()
+    var rand:Int = 0
+
+    if id == nil {
+        rand = numGenerator.random() % 10000 + 1
+    } else {
+        rand = Int(id!)!
+    }
+
+    let querySuccess = mysql.query(statement: "SELECT id, randomNumber FROM World WHERE id = \(rand)")
+
+    guard querySuccess else {
+
+        errorObject["id"] = "Failed to execute query"
+
+        return errorObject
+    }
+ 
+    let results = mysql.storeResults()!
+
+    results.forEachRow { row in
+
+        if let id = row[0], let randomNumber = row[1] {
+
+            returnObj["id"] = id
+            returnObj["randomNumber"] = randomNumber
+        } else {
+
+            returnObj["id"] = "No return value"
+            returnObj["randomNumber"] = "what happened?"
+        }
+    }
+
+    return returnObj
+}
+
+func updateOneFromWorld() -> [String: Any] {
+
+    var returnObj = [String: Any]()
+    var errorObject = [String: Any]()
+
+    let rand = numGenerator.random() % 10000
+    let rand2 = numGenerator.random() % 10000
+
+    let querySuccess = mysql.query(statement: "UPDATE World SET randomNumber = \(rand) WHERE id = \(rand2)")
+
+    guard querySuccess else {
+
+        errorObject["id"] = "Failed to execute query"
+
+        return errorObject
+    }
+ 
+    if let results = mysql.storeResults() {
+
+        results.forEachRow { row in
+
+            if let id = row[0], let randomNumber = row[1] {
+
+                returnObj["id"] = id
+                returnObj["randomNumber"] = randomNumber
+            } else {
+
+                returnObj["id"] = "No return value"
+                returnObj["randomNumber"] = "what happened?"
+            }
+        }
+
+        return returnObj
+    } else {
+
+        returnObj["id"] = rand2
+        returnObj["randomNumber"] = rand
+        return returnObj
+    }
+}
+
+func fortunesHandler(request: HTTPRequest, response: HTTPResponse) {
+
+    var arrOfFortunes = fetchFromFortune()
+
+    let newObj: [String: String] = ["id": "0", "message": "Additional fortune added at request time."]
+
+    arrOfFortunes.append(newObj)
+
+    let sortedArr = arrOfFortunes.sorted(by: ({ $0["message"]! < $1["message"]! }))
+
+    let htmlToRet = spoofHTML(fortunesArr: sortedArr)
+
+    response.appendBody(string: htmlToRet)
+    
+    setHeaders(response: response, contentType: "text/html")
+    response.setHeader(.custom(name: "CustomLength"), value: String(describing: htmlToRet.count + 32))
+
+    response.completed()
+}
+
+func updatesHandler(request: HTTPRequest, response: HTTPResponse) {
```

Comment:
I don't believe this meets the requirements for "Test type 5: Database updates":

> 4. The request handler must retrieve a set of World objects, equal in count to the `queries` parameter, from the World database table.

I don't see this happening anywhere in this request handler. It calls `updateOneFromWorld` _n_ times, but that function simply executes one `UPDATE` statement. There is no `SELECT` statement that I can see, particularly one that occurs before the DB is updated.

> 6. At least the `randomNumber` field must be read from the database result set.

The code on lines 138-143 fakes a `returnObj` _without_ reading it from the database. The presence of that code strongly suggests that `randomNumber` is not being read from the database result set. No such code should be necessary in a conformant implementation.

> 7. Each World object must have its `randomNumber` field updated to a new random integer between 1 and 10000.
> 8. Each World object must be persisted to the database with its new `randomNumber` value.

There does not appear to be any in-memory update of a "World" domain object that happens before the DB is updated. The update is a direct update of a DB row (on line 113) with an interpolated string, not a dynamically-generated query based on ORM-style persistence.